cni/pkg/taint/taint.go

```diff
@@ -0,0 +1,185 @@
+	return ts.configs
+}
+func NewTaintSetter(clientset client.Interface, options *Options) (ts *Setter, err error) {
+	configmap, err := clientset.CoreV1().ConfigMaps(options.ConfigmapNamespace).Get(context.TODO(), options.ConfigmapName, metav1.GetOptions{})
+	if err != nil {
+		return ts, err
+	}
+	ts = &Setter{
+		configs: []ConfigSettings{},
+		Client:  clientset,
+	}
+	ts.LoadConfig(*configmap)
+	return
+}
+
+// load corresponding configmap's critical labels and their namespace
+func (ts *Setter) LoadConfig(config v1.ConfigMap) {
+	log.Debugf("Loading configmap %s in %s", config.Name, config.Namespace)
+	ts.configs = make([]ConfigSettings, 0) //clear previous one
+	for key, value := range config.Data {
+		log.Debugf("Loading %s ", key)
+		var tempst []ConfigSettings
+		err := yaml.Unmarshal([]byte(value), &tempst)
+		if err != nil {
+			log.Fatalf("cannot unmarshal data : %v", err)
+		}
+		for _, elem := range tempst {
+			for _, selector := range strings.Split(elem.LabelSelector, ",") {
+				selector = strings.TrimSpace(selector)
+				_, err = metav1.ParseToLabelSelector(selector)
+				if err != nil {
+					log.Fatalf("illegal selector, %s", err.Error())
+				}
+				ts.configs = append(ts.configs, ConfigSettings{Namespace: elem.Namespace, Name: elem.Name, LabelSelector: selector})
+			}
+		}
+		log.Infof("successfully loaded %s", tempst)
+	}
+}
+
+//only pod with NodeReadiness Toleracnce with effect no schedule or
+//a generalized tolerance with noschedule effect can be considered
+func (ts *Setter) validTolerance(pod v1.Pod) bool {
+	for _, toleration := range pod.Spec.Tolerations {
+		if (toleration.Key == TaintName || toleration.Key == "") &&
+			toleration.Operator == v1.TolerationOpExists &&
+			toleration.Effect == v1.TaintEffectNoSchedule {
+			return true
+		}
+	}
+	return false
+}
+
+//check whether current node have readiness
+func (ts *Setter) HasReadinessTaint(node *v1.Node) bool {
+	ts.mutex.Lock()
+	defer ts.mutex.Unlock()
+	for _, taint := range node.Spec.Taints {
+		if taint.Key == TaintName && taint.Effect == v1.TaintEffectNoSchedule {
+			return true
+		}
+	}
+
+	return false
+}
+
+//assumption: order of taint is not important
+func (ts *Setter) RemoveReadinessTaint(node *v1.Node) error {
+	ts.mutex.RLock()
+	defer ts.mutex.RUnlock()
+	updatedTaint := deleteTaint(node.Spec.Taints, &v1.Taint{Key: TaintName, Effect: v1.TaintEffectNoSchedule})
+	node.Spec.Taints = updatedTaint
+	updatedNodeWithTaint, err := ts.Client.CoreV1().Nodes().Update(context.TODO(), node, metav1.UpdateOptions{})
+	if err != nil || updatedNodeWithTaint == nil {
+		return fmt.Errorf("failed to update node %v after adding taint: %v", node.Name, err)
+	}
+	log.Infof("Successfully removed taint on node %v", updatedNodeWithTaint.Name)
+	return nil
+}
+
+// taint node with specific taint name with effect of no schedule
+//do nothing if it already have the readiness taint
+func (ts *Setter) AddReadinessTaint(node *v1.Node) error {
+	ts.mutex.Lock()
```

Comment:
what is this mutex protecting?

Reply:
add taint to node spec is not thread safe, which is tested by go race tests, thus it is necessary to add lock in add taint operation